aria/modeling/service.py

```diff
@@ -0,0 +1,1529 @@
+                                            backreference='source_node')
+
+    @declared_attr
+    def inbound_relationships(cls):
+        return cls.one_to_many_relationship('relationship',
+                                            foreign_key='target_node_fk',
+                                            backreference='target_node')
+
+    @declared_attr
+    def plugins(cls):
+        return cls.many_to_many_relationship('plugin')
+
+    @declared_attr
+    def host(cls):
+        return cls.relationship_to_self('host_fk')
+
+    # region orchestration
+
+    runtime_properties = Column(modeling_types.Dict)
+    scaling_groups = Column(modeling_types.List)
+    state = Column(Text, nullable=False)
+    version = Column(Integer, default=1)
+
+    @declared_attr
+    def service_name(cls):
+        return association_proxy('service', 'name')
+
+    @property
+    def ip(self):
+        # TODO: totally broken
+        if not self.host_fk:
+            return None
+        host_node = self.host
+        if 'ip' in host_node.runtime_properties:  # pylint: disable=no-member
+            return host_node.runtime_properties['ip']  # pylint: disable=no-member
+        host_node = host_node.node_template  # pylint: disable=no-member
+        host_ip_property = host_node.properties.get('ip')
+        if host_ip_property:
+            return host_ip_property.value
+        return None
+
+    # endregion
+
+    # region foreign_keys
+
+    __private_fields__ = ['type_fk',
+                          'host_fk',
+                          'service_fk',
+                          'node_template_fk']
+
+    # Node many-to-one to Type
+    @declared_attr
+    def type_fk(cls):
+        return cls.foreign_key('type')
+
+    # Node one-to-one to Node
+    @declared_attr
+    def host_fk(cls):
+        return cls.foreign_key('node', nullable=True)
+
+    # Service one-to-many to Node
+    @declared_attr
+    def service_fk(cls):
+        return cls.foreign_key('service')
+
+    # Node many-to-one to NodeTemplate
+    @declared_attr
+    def node_template_fk(cls):
+        return cls.foreign_key('node_template', nullable=True)
+
+    # endregion
+
+    def satisfy_requirements(self, context):
+        node_template = self.node_template
+        satisfied = True
+        for requirement_template in node_template.requirement_templates:
+            # Find target template
+            target_node_template, target_node_capability = \
+                requirement_template.find_target(context, node_template)
+            if target_node_template is not None:
+                satisfied = self._satisfy_capability(context,
+                                                     target_node_capability,
+                                                     target_node_template,
+                                                     requirement_template)
+            else:
+                context.validation.report('requirement "{0}" of node "{1}" has no target node '
```

Comment:
there should be unified model for logging, it's weird that some log msgs, while other report it to the parser's context...